lock-api/src/main/java/com/palantir/lock/watch/LockWatchStateUpdater.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import com.google.common.collect.RangeSet;
+import com.palantir.lock.LockDescriptor;
+
+@NotThreadSafe
+public class LockWatchStateUpdater implements LockWatchEvent.Visitor<Void> {
+    private final RangeSet<LockDescriptor> watches;
+    private final Map<LockDescriptor, LockWatchInfo> lockWatchState;
+    private final Set<UUID> unmatchedOpenLocksEvents;
+
+    /**
+     * LockWatchStateUpdater is a LockWatchEvent visitor that updates the set of current lock watches, the state of
+     * specific locks, and tracks the information on encountered open locks events. It is the responsibility of the
+     * caller to visit lock watch events in the correct order as they occurred, without skipping events.
```

Comment:
By this, do we mean that we make a new updater if we fall out of the window and/or there is a leader election?

Reply:
Discussed offline: Usage is once per batch of updates, so that's ok - the calling class is responsible for invalidation in the election/too-much-lag cases.